packages/store/ts/config/experimental/parseTable.ts

```diff
@@ -0,0 +1,75 @@
+import { SchemaAbiType, isSchemaAbiType } from "@latticexyz/schema-type";
+import { isPlainObject } from "./common";
+import { ParseKeySchemaInput, ParseKeySchemaOutput, parseKeySchema } from "./parseKeySchema";
+import { ParseValueSchemaInput, ParseValueSchemaOutput, parseValueSchema } from "./parseValueSchema";
+import { assertExhaustive } from "@latticexyz/common/utils";
+import { tableIdToHex } from "@latticexyz/common";
+
+/** @internal */
+export type TableShapeInput = Readonly<{
+  namespace?: string;
+  keySchema?: ParseKeySchemaInput;
+  valueSchema: ParseValueSchemaInput;
+  offchainOnly?: boolean;
+}>;
+
+// TODO: add support for ParseValueSchemaInput instead of SchemaAbiType?
+//       requires an isValueSchemaInput helper that is distinct enough from isParseTableInputShape
+export type ParseTableInput = SchemaAbiType | TableShapeInput;
+
+export type ParseTableOutput<
+  defaultNamespace extends string,
+  name extends string,
+  input extends ParseTableInput
+> = input extends SchemaAbiType
+  ? ParseTableOutput<defaultNamespace, name, { valueSchema: input }>
+  : input extends TableShapeInput
+  ? Readonly<{
+      type: input["offchainOnly"] extends true ? "offchainTable" : "table";
+      namespace: input["namespace"] extends string ? input["namespace"] : defaultNamespace;
+      name: name;
+      tableId: `0x${string}`;
+      keySchema: ParseKeySchemaOutput<
+        input["keySchema"] extends ParseKeySchemaInput
+          ? input["keySchema"]
+          : never extends input["keySchema"]
+          ? undefined
+          : never
+      >;
+      valueSchema: ParseValueSchemaOutput<input["valueSchema"]>;
+    }>
+  : never;
+
+// TODO: is there a better way to check this aside from just looking at the shape/keys of the object?
+
+/** @internal */
+export const tableInputShapeKeys = ["namespace", "keySchema", "valueSchema", "offchainOnly"] as const;
+
+/** @internal */
+export function isTableShapeInput(input: unknown): input is TableShapeInput {
+  if (!isPlainObject(input)) return false;
+  if (Object.keys(input).some((key) => !tableInputShapeKeys.includes(key as (typeof tableInputShapeKeys)[number])))
```

Comment:
we do an `isPlainObject` check first because `Object.keys` can have some funky behavior for non-objects:
<img width="390" alt="image" src="https://github.com/latticexyz/mud/assets/508855/81d0b8f7-f970-4e1e-a854-17a41b37cf5b">